src/wallet/wallet_rpc_server.cpp

```diff
@@ -337,30 +337,48 @@ namespace tools
     if (!m_wallet) return not_open(er);
     try
     {
-      res.balance = m_wallet->balance(req.account_index);
-      res.unlocked_balance = m_wallet->unlocked_balance(req.account_index);
-      res.multisig_import_needed = m_wallet->multisig() && m_wallet->has_multisig_partial_key_images();
-      std::map<uint32_t, uint64_t> balance_per_subaddress = m_wallet->balance_per_subaddress(req.account_index);
-      std::map<uint32_t, uint64_t> unlocked_balance_per_subaddress = m_wallet->unlocked_balance_per_subaddress(req.account_index);
-      std::vector<tools::wallet2::transfer_details> transfers;
-      m_wallet->get_transfers(transfers);
-      std::set<uint32_t> address_indices = req.address_indices;
-      if (address_indices.empty())
-      {
-        for (const auto& i : balance_per_subaddress)
-          address_indices.insert(i.first);
-      }
-      for (uint32_t i : address_indices)
-      {
-        wallet_rpc::COMMAND_RPC_GET_BALANCE::per_subaddress_info info;
-        info.address_index = i;
-        cryptonote::subaddress_index index = {req.account_index, info.address_index};
-        info.address = m_wallet->get_subaddress_as_str(index);
-        info.balance = balance_per_subaddress[i];
-        info.unlocked_balance = unlocked_balance_per_subaddress[i];
-        info.label = m_wallet->get_subaddress_label(index);
-        info.num_unspent_outputs = std::count_if(transfers.begin(), transfers.end(), [&](const tools::wallet2::transfer_details& td) { return !td.m_spent && td.m_subaddr_index == index; });
-        res.per_subaddress.push_back(info);
+      uint32_t account_indices = req.account_index;
+      // helper loop counter variable
+      uint32_t account_index = account_indices;
+      if (req.all_accounts)
+      {
+        // if all_accounts was requested, get num of all accounts and iterate through them
+        account_indices = m_wallet->get_num_subaddress_accounts() - 1;
+        account_index = 0;
+      }
+      for (; account_index <= account_indices; account_index++)
+      {
+        res.balance = m_wallet->balance(account_index);
+        res.unlocked_balance = m_wallet->unlocked_balance(account_index);
+        res.multisig_import_needed = m_wallet->multisig() && m_wallet->has_multisig_partial_key_images();
```

Comment:
Ditto for both these

Reply:
fixed with new json response structure.

Reply:
Why a new response structure when you could have just used `+=` for these?